File: java/socketd/src/main/java/org/noear/socketd/transport/core/StreamManger.java
package org.noear.socketd.transport.core;

/**
 * 流管理器
 *
 * @author noear
 * @since 2.0
 */
public interface StreamManger {
    /**
     * 添加流
     *
     * @param sid    流Id
     * @param stream 流
     */
    void addStream(String sid, StreamInternal stream);

    /**
     * 获取流
     *
     * @param sid 流Id
     */
    StreamInternal getStream(String sid);

    /**
     * 移除流
     *
     * @param sid 流Id
     */
    void removeStream(String sid);
}

File: python/socketd/socketd/transport/core/stream/StreamManger.py
import asyncio
from abc import ABC, abstractmethod

from socketd.transport.core.stream.Stream import Stream


class StreamManger(ABC):

    @abstractmethod
    def add_stream(self, sid: str, stream: 'StreamInternal'): ...

    @abstractmethod
    def get_stream(self, sid: str) -> Stream: ...

    @abstractmethod
    def remove_stream(self, sid: str) -> None: ...


class StreamInternal(Stream, ABC):

    @abstractmethod
    def demands(self) -> int: ...

    @abstractmethod
    def timeout(self) -> float: ...

    @abstractmethod
    def insurance_start(self, streamManger: StreamManger, streamTimeout: float) -> None: ...

    @abstractmethod
    def insurance_cancel(self) -> None: ...

    @abstractmethod
    async def on_reply(self, reply) -> None | asyncio.Future: ...

    @abstractmethod
    def on_error(self, error): ...

    @abstractmethod
    def on_progress(self, is_send, val, max_val): ...
